cmd/daily-sweeper.go

```diff
@@ -66,23 +86,48 @@ func sweepRound(ctx context.Context, objAPI ObjectLayer) error {
 		return err
 	}
 
+	var interestedListeners []sweepListener
+
+	// Calculate the list of listeners that want
+	// to receive the list of buckets/objects.
+	for _, l := range copyDailySweepListeners() {
+		if l.IsRightTime() {
+			l.MarkStart()
+			interestedListeners = append(interestedListeners, l)
+		}
+	}
+
 	// List all objects, having read quorum or not in all buckets
 	// and send them to all the registered sweep listeners
 	for _, bucket := range buckets {
+		var perBucketListeners []sweepListener
+		for _, l := range interestedListeners {
+			if l.IsRightBucket(bucket.Name) {
+				perBucketListeners = append(perBucketListeners, l)
+			}
+		}
+
+		if len(perBucketListeners) == 0 {
+			// No listener for this bucket, move on
+			continue
+		}
+
 		// Send bucket names to all listeners
-		for _, l := range copyDailySweepListeners() {
-			l <- bucket.Name
+		for _, l := range perBucketListeners {
+			l.Send(sweepEntry{bucket: bucket})
```

Comment:
Is there a way we can ensure that the daily sweeper doesn't get indefinitely stuck if the channel inside the listener's implementation blocks for some reason?

Reply:
> Is there a way we can ensure that the daily sweeper doesn't get indefinitely stuck if the channel inside the listener's implementation blocks for some reason?

I thought about this, but actually we don't have too, it will never block, besides what timeout value can we pick ? like healing one 5 Tb object would probably take time in a not so fast machine